src/exec/graphics.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2022 System76 <info@system76.com>
+// SPDX-License-Identifier: GPL-3.0-only
+
+use std::collections::HashSet;
+use std::hash::{Hash, Hasher};
+use std::io;
+use std::ops::Deref;
+use std::path::PathBuf;
+
+#[derive(Debug, Default)]
+pub struct Gpus {
+    devices: Vec<Dev>,
+    default: Option<Dev>,
+}
+
+impl Gpus {
+    // Get gpus via udev
+    pub fn load() -> Self {
+        let drivers = get_gpus();
+
+        let mut gpus = Gpus::default();
+        for dev in drivers.unwrap() {
+            if dev.is_default {
+                gpus.default = Some(dev)
+            } else {
+                gpus.devices.push(dev)
+            }
+        }
+
+        gpus
+    }
+
+    /// `true` if there is at least one non default gpu
+    pub fn is_switchable(&self) -> bool {
+        self.default.is_some() && !self.devices.is_empty()
+    }
+
+    /// Return the default gpu
+    pub fn get_default(&self) -> Option<&Dev> {
+        self.default.as_ref()
+    }
+
+    /// Get the first non-default gpu, the current `PreferNonDefaultGpu` specification
+    /// Does not tell us which one should be used. Anyway most machine out there should have
+    /// only one discrete graphic card.
+    /// see: https://gitlab.freedesktop.org/xdg/xdg-specs/-/issues/59
+    pub fn non_default(&self) -> Option<&Dev> {
+        self.devices.first()
+    }
+}
+
+#[derive(Debug)]
+pub struct Dev {
+    id: usize,
+    driver: Driver,
+    is_default: bool,
+    parent_path: PathBuf,
+}
+
+impl Dev {
+    /// Get the environment variable to launch a program with the correct gpu settings
+    pub fn launch_options(&self) -> Vec<(String, String)> {
+        let dev_num = self.id.to_string();
+        let mut options = vec![];
+
+        match self.driver {
+            Driver::Unknown | Driver::Amd(_) | Driver::Intel => {
+                options.push(("DRI_PRIME".into(), dev_num))
+            }
+            Driver::Nvidia => {
+                options.push(("__GLX_VENDOR_LIBRARY_NAME".into(), "nvidia".into()));
+                options.push(("__NV_PRIME_RENDER_OFFLOAD".into(), "1".into()));
+                options.push((" __VK_LAYER_NV_optimus".into(), "NVIDIA_only".into()));
+            }
+        }
+
+        match self.get_vulkan_icd_paths() {
+            Ok(vulkan_icd_paths) if !vulkan_icd_paths.is_empty() => {
+                options.push(("VK_ICD_FILENAMES".into(), vulkan_icd_paths.join(":")))
+            }
+            Err(err) => eprintln!("Failed to open vulkan icd paths: {err}"),
+            _ => {}
+        }
+
+        options
+    }
+
+    // Lookup vulkan icd files and return the ones matching the driver in use
+    fn get_vulkan_icd_paths(&self) -> io::Result<Vec<String>> {
+        let vulkan_icd_paths = dirs::data_dir()
+            .expect("local data dir does not exists")
+            .join("vulkan/icd.d");
+        let vulkan_icd_paths = &[PathBuf::from("/usr/share/vulkan/icd.d"), vulkan_icd_paths];
```

Comment:
We might want to make this a compile-time env variable.

Reply:
You mean `XDG_DATA_DIRS` or the whole path to the the icd files ?

Reply:
For the `/usr/share/` path

Reply:
I don't get it, you mean a `const` or I am not aware of the env var holding this path ?

Reply:
Something like `std::env!("VULKAN_ICD_PATH", "must define system Vulkan ICD path")`

Reply:
It's a compile-time panic in case the person packaging software didn't define it.

Reply:
Ah sorry I was not aware :smile:

Reply:
Yeah I should be using it more often, myself.

Reply:
it's done